### PR TITLE
[New Event Representation]: Bina-Rep event frames

### PR DESCRIPTION
**Pull request by Barchid**
Hello,

I have added a new event representation: Bina-Rep, from the paper ([Bina-Rep Event Frames: a Simple and Effective Representation for Event-based cameras](https://arxiv.org/abs/2202.13662) . The paper is accepted at [ICIP 2022, in the special session dedicated to neuromorphic image acquisition & analysis](https://2022.ieeeicip.org/special-sessions/#) 

Briefly explained, it is a sort of "post-processing" method after running a `ToFrame()` representation. Let's say you obtain `T*N` binary frames after a `ToFrame()` transform. Bina-Rep decomposes this sequence of frames into `T` frames of `N-bit` numbers (which are named "bina-rep" frames). The advantage of bina-rep frames is to have an event representation that contains:
1. The presence of an event (*similar to simple frame representation*)
2. The number of accumulated events (*similar to event frame representation*)
3. The order of arrival of these events in the original `N` binary frames that are used to produce the bina-rep frame

I want to draw your attention about one trick I implemented for easier use. The transform dataclass (in `tonic/transforms.py`) is not frozen like the others. I removed this flag to implement a `__post_init__()` function that creates a default `ToFrame` transform used in the bina-rep code. As I don't know your requirements (and I'm afraid I am not an experienced Open-Source contributor), I wanted to know if this is ok for you.

Thanks for your hard work!

### CODE REVIEW

File: tonic/functional/to_bina_rep.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    https://arxiv.org/pdf/2202.13662.pdf
+    Parameters:
+        events: numpy.ndarray of shape [num_events, num_event_channels].
+        to_frame_transform (tonic.transforms.ToFrame): the ToFrame transform that creates the T*B binary event frames.
+        n_frames (int): the number T of bina-rep frames.
+        n_bits (int): the number N of bits used in the N-bit representation.
+    Returns:
+        (numpy.ndarray) the sequence of bina-rep event frames with dimensions (TxPxHxW).
+    """
+    assert "x" and "t" and "p" in events.dtype.names
```
Comment (fabhertz95) [inline tonic/functional/to_bina_rep.py:23]: Why is `y` not checked?

**Comment by fabhertz95** [comment]
Hey @Barchid, very nice! Basically, you arrange N events in a one-hot notation on N bits for each pixel, in which each bit index corresponds to the event order of arrival? It reminds me of HATS without the first-order filtering of the event neighbourhood and  using a `2**i` weight for each event in the accumulation.

File: tonic/transforms.py [hunk 2]
```diff
@@ -643,6 +643,35 @@ def __call__(self, events):
...
+    https://arxiv.org/pdf/2202.13662.pdf
+    Parameters:
+        to_frame_transform (tonic.transforms.ToFrame): the ToFrame transform that creates the T*B binary event frames. Defaults to None.
+        n_frames (int): the number T of bina-rep frames.
+        n_bits (int): the number N of bits used in the N-bit representation.
+    """
+
+    n_frames: Optional[int] = 1
+    n_bits: Optional[int] = 8
+    to_frame_transform: Optional[ToFrame] = None
```
Comment (biphasic) [inline tonic/transforms.py:660]: it's good practice to re-use existing code, but I have the feeling that here the user is given to many options. Do you think the user will change ToFrame's overlap and include_incomplete options? Because if not, then I suggest adding `sensor_size` as a parameter and to call ToFrame internally, without exposing it to the user. So basically
```
transform = transforms.ToBinaRep(
            sensor_size=sensor_size,
            n_frames=n_frames,
            n_bits=n_bits,
        )
```
That will also take care of the __postinit__. It's not a problem that the dataclass is not frozen.

File: tonic/functional/to_bina_rep.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    events: np.ndarray,
+    to_frame_transform,
+    n_frames: int = 1,
+    n_bits: int = 8,
+):
+    """Representation that takes T*B binary event frames to produce a sequence of T frames of N-bit numbers.
+    To do so, N binary frames are interpreted as a single frame of N-bit representation. Taken from the paper
+    Barchid et al. 2022, Bina-Rep Event Frames: a Simple and Effective Representation for Event-based cameras
+    https://arxiv.org/pdf/2202.13662.pdf
+    Parameters:
```
Comment (biphasic) [inline tonic/functional/to_bina_rep.py:15]: I'm not 100% sure but I think for the docs to render this properly it needs one line of whitespace above `Parameters:` and `Returns:`

**Review by biphasic — CHANGES_REQUESTED** [review]
Thank you @Barchid for you PR! This is a simple yet powerful representation. It seems quite expensive to compute and I'm wondering if you could create it from the raw events by slicing them in time and then creating the binary frames from N events directly. However I appreciate the code re-use of ToFrame, which makes it easier to maintain. For the moment for it not to be a considerable bottleneck during training people could make use of [disk-caching](https://tonic.readthedocs.io/en/latest/tutorials/fast_dataloading.html#dataset-caching) to speed up everything after the first epoch. We can optimize it later on if need be.
Could you please add your transform to the docs folder as well? https://github.com/neuromorphs/tonic/blob/develop/docs/reference/transformations.rst Thank you

**Comment by biphasic** [comment]
Also I have another indirectly related question. For you experiments in the paper, did you use Tonic's ToVoxelGrid transform? If not, could you share the differences to your implementation in a separate discussion please? For example [here](https://github.com/neuromorphs/tonic/discussions)

File: tonic/functional/to_bina_rep.py (hunk 1 above)
Comment (Barchid) [reply tonic/functional/to_bina_rep.py:23]: I took this assert from the `to_frame_numpy()` function, that works with 1-dim input like audio record (if I understood well). I thought I could reuse this ability of `ToFrame` in my implementation, but I just figured out that my code cannot adapt to 1-D events. I will add the assert in the next commit.

File: tonic/transforms.py (hunk 2 above)
Comment (Barchid) [reply tonic/transforms.py:660]: The choice of allowing the user to choose the ToFrame parameters is mostly inspired from my experience (for the paper and for future works). I often finetune the parameters of `ToFrame()` depending on the DVS records I work with (e.g. the record of a table tennis match is a real pain when working with `n_time_bins`, as it creates event frames that can vary a lot in appearance when the ball is smashed or just sliced). That's why I felt that it was more usable to make the `ToFrame` processing completely exposed to fit the user's needs.

But as I am one of the author, I may not be aware of how wide the options are for a user that just wants to check how the transform works. If you are sure that closing `ToFrame` is a better option for user-friendliness, I can change it.

File: tonic/functional/to_bina_rep.py (hunk 3 above)
Comment (Barchid) [reply tonic/functional/to_bina_rep.py:15]: You are completely right. I'm going to fix that.

File: tonic/transforms.py (hunk 2 above)
Comment (biphasic) [reply tonic/transforms.py:660]: I see, that makes sense. Let's keep it flexible! However, I'm still not a fan of how the frame transform is passed in a constructor. Reason for this is that I'd like to keep the API fairly consistent where possible. If a transform only works on a specific kind of data (raw events, frames,...), then normally the user has to chain the transforms. How about the following: 
* We call BinRep a 'transformation that works on frames', which only accepts as data numpy arrays of shape (C, H, W). 
* The transform chain would then look something like this:
```
import tonic.transforms as transforms

transform = transforms.Compose([
    transforms.ToFrame(
        sensor_size=sensor_size,
        n_time_bins=n_time_bins,
    ),
    transforms.ToBinaRep(
        n_frames=n_frames,
        n_bits=n_bits,
    ),
])
```
I feel like that would make it very clear to a user that BinaRep works on frames and reflects the general way of how to chain transforms in Tonic. How does that sound to you?

Comment (Barchid) [reply tonic/transforms.py:660]: Sounds good to me! I will update the code and documentation.

Comment (biphasic) [reply tonic/transforms.py:660]: Great, thank you. One last thing. To make it really easy for the user, could you also include an 'example' in the docstring much like here: https://github.com/neuromorphs/tonic/blob/develop/tonic/transforms.py#L14
You could take the example I posted above that includes the separate ToFrame transform. Thank you!

**Comment by Barchid** [comment]
Thanks @biphasic  and @fabhertz95  for your feedbacks and questions. I integrated your change requests and fixed the problems in the documentation. As for your questions, I will answer them below.

> Also I have another indirectly related question. For you experiments in the paper, did you use Tonic's ToVoxelGrid transform? If not, could you share the differences to your implementation in a separate discussion please? For example [here](https://github.com/neuromorphs/tonic/discussions)

Yes, we used the `ToVoxelGrid` transform from `tonic`. Honestly, it saved us a good amount of time to have such method at hand, btw :)

> Hey @Barchid, very nice! Basically, you arrange N events in a one-hot notation on N bits for each pixel, in which each bit index corresponds to the event order of arrival? It reminds me of HATS without the first-order filtering of the event neighbourhood and using a `2**i` weight for each event in the accumulation.

This is a good observation. We have not seen it like this but you surely showed similarities between the two representations. It would be interesting to investigate how far these similarities go, both in theory and in practical use.

File: tonic/transforms.py (hunk 2 above)
Comment (Barchid) [reply tonic/transforms.py]: Done.

**Comment by biphasic** [comment]
this looks very good now. Merging
